app/src/main/java/com/example/sharingang/MapActivity.kt

```diff
@@ -1,62 +1,156 @@
     private lateinit var binding: ActivityMapBinding
-    private var lastLocation: Location? = null
+
     private lateinit var fusedLocationClient: FusedLocationProviderClient
+    private var permissionGranted: Boolean = false
 
-    // Allows the cancellation of a location request if, for example, the user exists the activity
-    private var cancellationTokenSource = CancellationTokenSource()
+    private val requestPermissionLauncher =
+        registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted: Boolean ->
+            if (isGranted) {
+                permissionGranted = true
+                startLocationUpdates()
+            } else {
+                Toast.makeText(
+                    this,
+                    "Location permission successfully denied. Feature is disabled.",
+                    Toast.LENGTH_LONG
+                ).show()
+                permissionGranted = false
+            }
+        }
+
+    private var lastLocation: Location? = null
+    private lateinit var locationCallback: LocationCallback
+    private var lastLocationMarker: Marker? = null
+    private var map: GoogleMap? = null
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         binding = DataBindingUtil.setContentView(this, R.layout.activity_map)
         fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
-        binding.buttonGetLocation.setOnClickListener {
-            checkLocationPermissions()
+
+        locationCallback = object : LocationCallback() {
+            override fun onLocationResult(locationResult: LocationResult) {
+                lastLocation = locationResult.lastLocation
+                updateLocationText()
+                moveCameraToLastLocation()
+                moveLastLocationMarker()
+            }
         }
+        binding.mapView.onCreate(savedInstanceState)
+        binding.mapView.getMapAsync(this)
     }
 
+
     @SuppressLint("MissingPermission")
-    private fun updateLocation() {
-        val currentLocationTask: Task<Location> = fusedLocationClient.getCurrentLocation(
-                LocationRequest.PRIORITY_HIGH_ACCURACY,
-                cancellationTokenSource.token
-        )
-        currentLocationTask.addOnCompleteListener { task: Task<Location> ->
-            if (task.isSuccessful && task.result != null) {
-                lastLocation = task.result
-                binding.locationDisplay.text = String.format("Your location is: %s %s", lastLocation!!.longitude, lastLocation!!.latitude)
-            }
+    private fun startLocationUpdates() {
+        if (!permissionGranted) {
+            requestLocationPermission()
+        } else {
+            fusedLocationClient.requestLocationUpdates(
+                LocationRequest.create().apply {
+                    interval = 5000
+                    fastestInterval = 2500
+                    priority = LocationRequest.PRIORITY_HIGH_ACCURACY
```

Comment:
What is the difference between `interval` and `fastestInterval`? How do we choose which interval we are going to be using for updating the location?